dvc/commands/dag.py

```diff
@@ -133,6 +171,225 @@ def _build(repo, target=None, full=False, outs=False):
+                "clean": resource_clean,
+                "pushed": resource_pushed,
+            }
+
+        pipeline.nodes[stage.addressing][rsc_key + " clean"] = all_resources_clean
+        pipeline.nodes[stage.addressing][rsc_key + " pushed"] = all_resources_pushed
+
+
+def _invalidate_downstream(pipeline: "DiGraph", node: Hashable) -> None:
+    """
+    Invalidates the downstream nodes of the given node in the pipeline
+    by setting their 'validated' attribute to False.
+
+    Args:
+        pipeline: The pipeline graph containing the node.
+        node: The node whose downstream nodes are to be invalidated.
+
+    Returns:
+        None.
+    """
+    # Don't invalidate frozen nodes
+    if not pipeline.nodes[node]["frozen"]:
+        # Set the 'validated' attribute to False for the current node
+        pipeline.nodes[node]["validated"] = False
+
+        # Recursively invalidate downstream nodes
+        for successor_id in pipeline.successors(node):
+            _invalidate_downstream(pipeline, successor_id)
+
+
+def _validate_pipeline(pipeline: "DiGraph") -> None:
+    """
+    Validates the nodes in the pipeline by setting their 'validated'
+    attribute based on whether they have changes or upstream changes.
+
+    Args:
+        pipeline: The pipeline graph containing the nodes.
+
+    Returns:
+        None.
+    """
+    # pipeline is right now a dependency graph, we reverse it to make it a
+    # "data flow" graph. Then we can invalidate "downstream" stages if a
+    # stage is not green.
+    pipeline.reverse()
+
+    # Set the 'validated' attribute to True for all nodes by default
+    for node in pipeline.nodes:
+        pipeline.nodes[node]["validated"] = True
+
+    for node in pipeline.nodes:
+        node_data = pipeline.nodes[node]
+
+        # Check if any attribute has a False value
+        if not all(node_data.get(key, True) for key in VALIDATION_ATTRIBUTES):
+            _invalidate_downstream(pipeline, node)
+
+    # reverse the reversion
+    pipeline.reverse()
+
+
+def _set_stage_status(pipeline: "DiGraph"):
+    """
+    Sets the 'status' attribute for each stage node in the pipeline,
+    summarizing its status for rendering in a DAG.
+
+    Args:
+        pipeline: The pipeline graph containing the stage nodes.
+
+    Returns:
+        None.
+    """
+    for _, data in pipeline.nodes(data=True):
+        command_run = data.get("command run", None)
+        outs_clean = data.get("outs clean", None)
+        outs_pushed = data.get("outs pushed", None)
+        deps_clean = data.get("deps clean", None)
+        deps_pushed = data.get("deps pushed", None)
+        validated = data.get("validated", None)
+        frozen = data.get("frozen", False)
+
+        # Stage is not validated but otherwise green -> grey
+        if validated is False and all(
+            [outs_clean, outs_pushed, deps_clean, deps_pushed, command_run]
+        ):
+            data["status"] = "grey"
```

Comment:
It would be better to use explicit `status` names instead of color names and let the downstream representation map `status` to color or whatever  (i.e. `_get_class_defs` in mermaid).